taichi/backends/metal/codegen_metal.cpp

```diff
@@ -110,6 +124,28 @@ class RootIdsExtractor : public BasicStmtVisitor {
   std::unordered_set<int> roots_;
 };
 
+class TaskPreprocessor final : public BasicStmtVisitor {
```

Comment:
Should we collect this into a analysis pass? (i.e. dump out usage and capability information, like random, numerical types used, SNode types used, external arrays, etc.)

Reply:
+1, this is very useful!